misk/src/main/kotlin/misk/CoordinatedService2.kt

```diff
@@ -0,0 +1,238 @@
+package misk
+
+import com.google.common.util.concurrent.AbstractService
+import com.google.common.util.concurrent.MoreExecutors
+import com.google.common.util.concurrent.Service
+import com.google.common.util.concurrent.Service.Listener
+import com.google.common.util.concurrent.Service.State
+
+/**
+ * Services in Misk can depend on other services.
+ *
+ * ### Dependencies
+ *
+ * Suppose we have a DatabaseService and a MovieService, with the MovieService depending on the
+ * DatabaseService.
+ *
+ * ```
+ * DatabaseService
+ *   depended on by MovieService
+ * ```
+ *
+ * This class manages startup and shutdown of the each service, so that a service can only run when
+ * the services it depends on are running. In the example above, the MovieService doesn't enter the
+ * `STARTING` state until the DatabaseService has entered the `RUNNING` state. Conversely, the
+ * MovieService must enter the `TERMINATED` state before the DatabaseService enters the `STOPPING`
+ * state.
+ *
+ * Dependencies can have their own dependencies, so there's an entire graph to manage of what starts
+ * and stops when.
+ *
+ * ### Enhancements
+ *
+ * Some services exist to enhance the behavior of another service. For example, a DatabaseService
+ * may manage a generic connection to a MySQL database, and the SchemaMigrationService may create
+ * tables specific to the application.
+ *
+ * We treat such enhancements as implementation details of the enhanced service: they depend on the
+ * service, but downstream dependencies like the MovieService don't need to know that they exist.
+ *
+ * ```
+ * DatabaseService
+ *   enhanced by SchemaMigrationService
+ *   depended on by MovieService
+ * ```
+ *
+ * In the above service graph we start the DatabaseService first, the SchemaMigrationService second,
+ * and finally the MovieService. The MovieService doesn't need to express a dependency on the
+ * SchemaMigrationService, that happens automatically for enhancements.
+ *
+ * ### How It Works
+ *
+ * CoordinatedService2 wraps the actual Service to defer its start up and shut down until its
+ * dependent services are ready.
+ *
+ * This Service will stall in the `STARTING` state until all upstream services are `RUNNING`.
+ * Symmetrically it stalls in the `STOPPING` state until all dependent services are `TERMINATED`.
+ */
+internal class CoordinatedService2(val service: Service) : AbstractService() {
```

Comment:
Is there going to be fast-follow PR replacing of all uses of the existing ``CoordinatedService``?

Reply:
I'm still new to the project and this PR was made with a lot of guidance from @swankjesse... But I believe that's the idea, yes.

Reply:
Yep. Kill CoordinatedService, then take its name.